src/ray/gcs/gcs_server/node_info_handler_impl.h

```diff
@@ -57,8 +60,25 @@ class DefaultNodeInfoHandler : public rpc::NodeInfoHandler {
                              SendReplyCallback send_reply_callback) override;
 
  private:
+  void UnregisterNode(const ClientID &node_id, rpc::GcsNodeInfo &node_info,
+                      rpc::UnregisterNodeReply *reply,
+                      const SendReplyCallback &send_reply_callback);
+
+  void DeleteResources(const ClientID &node_id,
+                       const std::vector<std::string> &resource_names,
+                       const gcs::NodeInfoAccessor::ResourceMap &delete_resources,
+                       DeleteResourcesReply *reply,
+                       const SendReplyCallback &send_reply_callback);
+
   gcs::RedisGcsClient &gcs_client_;
   gcs::GcsNodeManager &gcs_node_manager_;
+  gcs::GcsPubSub gcs_pub_sub_;
+
+  /// Mutex to protect the nodes_cache_ field.
+  absl::Mutex node_resource_mutex_;
+  /// A mapping from node id to node resource.
+  std::unordered_map<ClientID, gcs::NodeInfoAccessor::ResourceMap> nodes_resource_cache_
```

Comment:
I think it is better to add a `GcsResourceManager`` to handle the resource related logic.

Reply:
fixed